kotlinx-collections-immutable/src/main/kotlin/kotlinx/collections/immutable/implementations/immutableList/PersistentVector.kt

```diff
@@ -117,120 +139,131 @@ internal class PersistentVector<E>(private val root: Array<Any?>,
-            newRoot[bufferIndex] = null
-            return newRoot
+        val newBufferAtIndex = if (shift == LOG_MAX_BUFFER_SIZE) {
+            tailCarry.value = root[bufferIndex]
+            null
+        } else {
+            @Suppress("UNCHECKED_CAST")
+            pullLastBuffer(root[bufferIndex] as Array<Any?>, shift - LOG_MAX_BUFFER_SIZE, index, tailCarry)
         }
-        val bufferAtIndex = pullLastBuffer(root[bufferIndex] as Array<Any?>, shift - LOG_MAX_BUFFER_SIZE, index, tailWrapper)
-        if (bufferAtIndex == null && bufferIndex == 0) {
+
+        if (newBufferAtIndex == null && bufferIndex == 0) {
             return null
         }
+
         val newRoot = root.copyOf(MAX_BUFFER_SIZE)
-        newRoot[bufferIndex] = bufferAtIndex
+        newRoot[bufferIndex] = newBufferAtIndex
         return newRoot
     }
 
-    private fun removeFromTail(root: Array<Any?>, rootSize: Int, shift: Int, index: Int): PersistentList<E> {
-        val tailFilledSize = size - rootSize
-        assert(index < tailFilledSize)
-
-        if (tailFilledSize == 1) {
-            return pullLastBufferFromRoot(root, rootSize, shift)
-        }
-        val newTail = tail.copyOf(MAX_BUFFER_SIZE)
-        if (index < tailFilledSize - 1) {
-            System.arraycopy(tail, index + 1, newTail, index, tailFilledSize - index - 1)
-        }
-        newTail[tailFilledSize - 1] = null
-        return PersistentVector(root, newTail, rootSize + tailFilledSize - 1, shift)
-    }
-
-    private fun removeFromRoot(root: Array<Any?>, shift: Int, index: Int, tailWrapper: ObjectWrapper): Array<Any?> {
-        val bufferIndex = (index shr shift) and MAX_BUFFER_SIZE_MINUS_ONE
+    /**
+     * Removes element from trie at the the specified [index].
+     *
+     * [tailCarry] on input contains the first element of the adjacent trie to fill the last vacant element with.
+     * [tailCarry] on output contains the first element of this trie.
+     *
+     * @return the new root of the trie.
+     */
+    private fun removeFromRootAt(root: Array<Any?>, shift: Int, index: Int, tailCarry: ObjectWrapper): Array<Any?> {
+        val bufferIndex = indexSegment(index, shift)
 
         if (shift == 0) {
             val newRoot = if (bufferIndex == 0) arrayOfNulls<Any?>(MAX_BUFFER_SIZE) else root.copyOf(MAX_BUFFER_SIZE)
-            System.arraycopy(root, bufferIndex + 1, newRoot, bufferIndex, MAX_BUFFER_SIZE - bufferIndex - 1)
-            newRoot[MAX_BUFFER_SIZE - 1] = tailWrapper.value
-            tailWrapper.value = root[0]
+            root.copyInto(newRoot, bufferIndex, bufferIndex + 1, MAX_BUFFER_SIZE)
+            newRoot[MAX_BUFFER_SIZE - 1] = tailCarry.value
+            tailCarry.value = root[0]
             return newRoot
         }
 
+        // Could be val bufferLastIndex = root.indexOfLast { it != null }, but that isn't optimized enough
         var bufferLastIndex = MAX_BUFFER_SIZE_MINUS_ONE
         while (root[bufferLastIndex] == null) {
```

Comment:
can be optimised to:
```
var bufferLastIndex = indexSegment(rootSize() - 1, shift)
if (bufferLastIndex < MAX_BUFFER_SIZE - 1 && root[bufferLastIndex + 1] != null) {
    bufferLastIndex = MAX_BUFFER_SIZE - 1
}
```

Reply:
ok, please apply these two loop-removal optimizations in a separate commit after this PR.